keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots/README.md

```diff
@@ -0,0 +1,1105 @@
+      name: bar
+      namespace: prod
+    spec:
+      from:
+      - group: snapshot.storage.k8s.io/v1alpha1
+        kind: VolumeSnapshotLink
+        namespace: test
+      to:
+      - group: snapshot.storage.k8s.io/v1
+        kind: VolumeSnapshot
+        name: foo-backup
+    ```
+2. In the test namespace, Bob creates a `VolumeSnapshotLink` foo-link that references the `VolumeSnapshot` foo-backup in the prod namespace as a source,
+    ```yaml
+    apiVersion: snapshot.storage.k8s.io/v1alpha1
+    kind: VolumeSnapshotLink
+    metadata:
+      name: foo-link
+      namespace: test
+    spec:
+      source:
+        name: foo-backup
+        namespace: prod
+    ```
+3. In the test namespace, Bob creates a `PersistentVolumeClaim` foo-testing that references the `VolumeSnapshotLink` foo-link as a data source,
+    ```yaml
+    apiVersion: v1
+    kind: PersistentVolumeClaim
+    metadata:
+      name: foo-testing
+      namespace: test
+    spec:
+      accessModes:
+      - ReadWriteOnce
+      resources:
+        requests:
+          storage: 10Mi
+      dataSourceRef:
+        apiGroup: snapshot.storage.k8s.io/v1alpha1
+        kind: VolumeSnapshotLink
+        name: foo-link
+      volumeMode: Filesystem
+    ```
+4. Once the populator finds a `VolumeSnapshotLink` is specified as `dataSourceRef`, it checks all `ReferenceGrants` in `VolumeSnapshotLink.spec.source.namespace` to see if populating the `VolumeSnapshotLink.spec.source` is allowed. If it is allowed, the populator populates the volume.
+    Note that how `ReferenceGrant` is checked depends on the implementation, however controllers that are trying to use the `VolumeSnapshot` in another namespace must check `ReferenceGrant` if the access is allowed before it actually starts exposing the data and metadata from the `VolumeSnapshot` to the `VolumeSnapshotLink`'s namespace.
+
+### API
+
+A new `VolumeSnapshotLink` CRD is introduced in `snapshot.storage.k8s.io` API group:
+
+```golang
+type VolumeSnapshotLink struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Spec VolumeSnapshotLinkSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+}
+
+type VolumeSnapshotLinkList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items []VolumeSnapshotLink `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// VolumeSnapshotLinkSpec describes attributes of a volume snapshot link.
+type VolumeSnapshotLinkSpec struct {
+	// This field is immutable after creation.
+	Source VolumeSnapshotLinkSource `json:"source" protobuf:"bytes,1,opt,name=source"`
+}
+
+// VolumeSnapshotLinkSource specifies a reference to VolumeSnapshot.
+type VolumeSnapshotLinkSource struct {
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Namespace string `json:"namespace" protobuf:"bytes,2,opt,name=namespace"`
```

Comment:
is it allowed to link to a snapshot in my own NS?  Does it still requiore a grant?

Reply:
> is it allowed to link to a snapshot in my own NS? Does it still requiore a grant?

@thockin Updated the KEP to allow not specifying namespace in `VolumeSnapshotLink` and provisioning from it without `ReferenceGrant`.